Sources/SwiftProtobuf/Decoder.swift

```diff
@@ -0,0 +1,132 @@
+// Sources/SwiftProtobuf/Decoder.swift - Basic field setting
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/master/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+///
+/// In this way, the generated code only knows about schema
+/// information; the decoder logic knows how to decode particular
+/// wire types based on that information.
+///
+// -----------------------------------------------------------------------------
+
+import Swift
+import Foundation
+
+/// This is the abstract protocol used by the generated code
+/// to deserialize data.
+///
+/// The generated code looks roughly like this:
+///
+/// ```
+///   while fieldNumber = try decoder.nextFieldNumber() {
+///      switch fieldNumber {
+///      case 1: decoder.decodeRepeatedInt32Field(value: &_field)
+///      ... etc ...
+///   }
+/// ```
+///
+/// For performance, this is mostly broken out into a separate method
+/// for singular/repeated fields of every supported type.  Note that
+/// we don't distinguish "packed" here, since all existing decoders
+/// treat "packed" the same as "repeated" at this level.  (That is,
+/// even when the serializer distinguishes packed and non-packed
+/// forms, the deserializer always accepts both.)
+///
+/// Generics come into play at only a few points: Enums and Messages
+/// use a generic type to locate the correct initializer.  Maps and
+/// Extensions use generics to avoid the method explosion of having to
+/// support a separate method for every map and extension type.  Maps
+/// do distinguish Enum-valued and Message-valued maps to avoid
+/// polluting the generated Enum and Message types with all of the
+/// necessary generic methods to support this.
+
+public protocol Decoder {
+    // Called by a Oneof when it already has a value and is being
+    // asked to accept a new value.  Some decoders specify that oneof
+    // decoding must fail in this case.  Those decoders throw a
+    // suitable error from this function; other decoders make it a
+    // no-op.  No existing decoder actually needs this to be mutating,
+    // but someone might want to track OneOf conflicts in the future.
+    mutating func handleConflictingOneOf() throws
+
+    // Get the next field number.  For JSON and Text, the decoder
+    // translates name to number at this point, based on information
+    // it obtained from the message when it was initialized.
+    mutating func nextFieldNumber() throws -> Int?
+
+    // Primitive field decoders
+    mutating func decodeSingularFloatField(value: inout Float) throws
```

Comment:
I don't think the `FieldType` types can go away entirely unless we commit to exporting a separate type for every possible Map and Extension.  But I do think all of the `FooTypeAddition.swift` files will disappear by the time this is done.